test/firtool/file-list.mlir

```diff
@@ -0,0 +1,63 @@
+// RUN: firtool %s --format=mlir --split-verilog -o=%t --blackbox-resource-path=%S/..
+// RUN: FileCheck %s --check-prefix=VERILOG-TOP < %t/test_mod.v
+// RUN: FileCheck %s --check-prefix=VERILOG-FOO < %t/magic/file-list-inline.v
+// RUN: FileCheck %s --check-prefix=VERILOG-HDR < %t/magic/file-list-inline.svh
+// RUN: FileCheck %s --check-prefix=VERILOG-BAR < %t/magic/file-list-resource.v
+// RUN: FileCheck %s --check-prefix=VERILOG-GIB < %t/magic/file-list-path.v
+// RUN: FileCheck %s --check-prefix=LIST-TOP < %t/filelist.f
+// RUN: FileCheck %s --check-prefix=LIST-BLACK-BOX < %t/magic.f
+
+// LIST-TOP: test_mod.v
+
+// LIST-BLACK-BOX:      magic/file-list-inline.v
+// LIST-BLACK-BOX-NEXT: magic/file-list-resource.v
+// LIST-BLACK-BOX-NEXT: magic/file-list-path.v
+
+firrtl.circuit "test_mod" attributes {annotations = [
+  // Black box processing should honor only the last annotation.
+  {class = "firrtl.transforms.BlackBoxTargetDirAnno", targetDir = "ignore_me_plz"},
+  {class = "firrtl.transforms.BlackBoxTargetDirAnno", targetDir = "magic"},
+  {class = "firrtl.transforms.BlackBoxResourceFileNameAnno", resourceFileName = "definitely_bad.f"},
+  {class = "firrtl.transforms.BlackBoxResourceFileNameAnno", resourceFileName = "magic.f"}
+]} {
+  // VERILOG-TOP-LABEL: module test_mod
+  // VERILOG-TOP-NEXT:    ExtInline foo
+  // VERILOG-TOP-NEXT:    ExtResource bar
+  // VERILOG-TOP-NEXT:    ExtPath gib
+  // VERILOG-TOP-NEXT:  endmodule
+  firrtl.module @test_mod() {
+    firrtl.instance @ExtInline {name = "foo", portNames = []}
+    firrtl.instance @ExtResource {name = "bar", portNames = []}
+    firrtl.instance @ExtPath {name = "gib", portNames = []}
+  }
+
+  // VERILOG-FOO-LABEL: module ExtInline(); endmodule
+  // VERILOG-HDR-LABEL: `define SOME_MACRO
+  firrtl.extmodule @ExtInline() attributes {annotations = [
```

Comment:
Firrtl really makes a single module to hold multiple blackboxs?

Reply:
The blackbox source code is annotated by the programmer in Chisel, and they're free to add multiple such annotations. There's some special casing of `*.svh` even that suggests that multiple annotations does occur in practice.

Reply:
It seems hacky to allow multiple files be generated for a single extmodule with a big attribute list.  When we lower to RTL, we should fix that, but if multi-file attributes exist in firrtl, we might as well preserver that in the firrtl dialect.

Reply:
I agree that it's not very elegant. But I could totally see it being necessary for some cases, e.g., if you have a IP-vendor-provided black box source file with a ``` `include "foo.svh" ```, forcing you to provide some config in `foo.svh` as well. 

@seldridge: Is this feature used in Chisel sources and should we preserve it?

Actually on second thought: Disallowing multiple file annotations would require adding more code to explicitly prevent a technically legitimate use case (and does not simplify downstream processing as far as I can see). Wouldn't that be a somewhat arbitrary restriction on the FIRRTL dialect with no technical merit?

Reply:
Yes, this should support multiple resources. E.g.: https://github.com/chipsalliance/rocket-chip/blob/147bdcc4a26c74e5d7a47e3d667d456699d6d11f/src/main/scala/devices/debug/Periphery.scala#L204. 

This is more a kludge of "add this file to the output directory" than it is about specifying the implementation of the blackbox. The implementation may be one file or multiple files. This borders on build system stuff.

Reply:
Since we have use cases for this in Chisel, it's probably easier to just support it and then make a conscious effort to phase things like simulation source file inclusion out upstream. Otherwise users get deterred by warnings like "you use multiple black box source files here".